rust/src/client.rs

```diff
@@ -0,0 +1,214 @@
+use crate::crypto::ByteObject;
+use crate::service::Handle;
+use crate::coordinator::RoundParameters;
+use crate::participant::{Participant, Task};
+use crate::{CoordinatorPublicKey, SumDict, PetError, InitError};
+use crate::UpdateSeedDict;
+use std::sync::Arc;
+use tokio::time;
+use std::time::Duration;
+use thiserror::Error;
+
+
+/// Client-side errors
+#[derive(Debug, Error)]
+pub enum ClientError {
+    /// Error starting the underlying [`Participant`]
+    #[error("failed to initialise participant: {0}")]
+    ParticipantInitErr(InitError),
+
+    /// Error from the underlying [`Participant`]
+    #[error("error arising from participant")]
+    ParticipantErr(PetError),
+
+    /// Error deserialising service data
+    #[error("failed to deserialise service data: {0}")]
+    DeserialiseErr(bincode::Error),
+
+    /// General client errors
+    #[error("unexpected client error")]
+    GeneralErr,
+}
+
+/// A client of the federated learning service
+///
+/// [`Client`] is responsible for communicating with the service, deserialising
+/// its messages and delegating their processing to the underlying
+/// [`Participant`].
+pub struct Client {
+    /// Handle to the federated learning [`Service`]
+    handle: Handle,
+
+    /// The underlying [`Participant`]
+    participant: Participant,
+
+    /// Interval to poll for service data
+    interval: time::Interval,
+
+    id: u32, // NOTE identifier for client for testing; may remove later
```

Comment:
So this is just for the logs right now?

Reply:
yes, they were introduced to aid traceability but don't serve any other purpose at the moment